storage/innobase/fil/fil0crypt.cc

```diff
@@ -1074,7 +1074,52 @@ static void fil_crypt_write_crypt_data_to_page0(fil_space_t *space) {
   mtr.commit();
 }
 
-bool fil_crypt_exclude_tablespace_from_rotation(fil_space_t *space) {
+bool fil_crypt_exclude_tablespace_from_rotation_temporarily(
+    fil_space_t *space) {
+  // We acquire fil_crypt_threads_mutex to stop encryption threads from
+  // generating crypt_data for tablespaces that they are about to encrypt.
+  // Generating crypt_data is the fist step in encryption process.
+  mutex_enter(&fil_crypt_threads_mutex);
+
+  if (space->crypt_data == nullptr) {
+    space->exclude_from_rotation = true;
+    mutex_exit(&fil_crypt_threads_mutex);
+    return true;
+  }
+
+  mutex_exit(&fil_crypt_threads_mutex);
+
+  // crypt_data already existed.
+  fil_space_crypt_t *crypt_data = space->crypt_data;
+  // take a lock on crypt_data to block rotation from starting
+  IB_mutex_guard crypt_data_mutex_guard(&crypt_data->mutex);
+
+  if (space->exclude_from_rotation) {
```

Comment:
Does this check need the crypt_data mutex?

Reply:
no, that's why I have made the exclude_from_rotation atomic operation - so it can be accessed safely with and without crypt_data mutex.

Reply:
In this case, can't we perform this check at the start of this function, before acquiring any mutexes?

Reply:
right, very much so, changing. Thanks :+1: